cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+        return intersection(r2, r1, cmp);
+    }
+    if (!starts_before_end(r2, r1, cmp)) {
+        return {};
+    }
+    const auto& intersection_start = r2.start();
+    const auto& intersection_end = ends_before_end(r1, r2, cmp) ? r1.end() : r2.end();
+    if (intersection_start == intersection_end && intersection_end.has_value()) {
+        return query::clustering_range::make_singular(intersection_end->value());
+    }
+    return query::clustering_range(intersection_start, intersection_end);
+}
+
+/// An expression visitor that translates multi-column atoms into clustering ranges.
+struct multi_column_range_accumulator {
+    const query_options& options;
+    const schema_ptr schema;
+    std::vector<query::clustering_range> ranges{query::clustering_range::make_open_ended_both_sides()};
+    const clustering_key_prefix::prefix_equal_tri_compare prefix3cmp = get_unreversed_tri_compare(*schema);
+
+    void operator()(const binary_operator& binop) {
+        if (auto tup = dynamic_pointer_cast<tuples::value>(binop.rhs->bind(options))) {
+            if (!is_compare(binop.op)) {
+                on_internal_error(
+                        rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
+            }
+            auto opt_values = tup->get_elements();
+            auto& lhs = std::get<std::vector<column_value>>(binop.lhs);
+            std::vector<bytes> values(lhs.size());
+            for (size_t i = 0; i < lhs.size(); ++i) {
+                values[i] = deref_column_value(opt_values[i], lhs.at(i).col->name_as_text());
+            }
+            intersect_all(to_range(binop.op, clustering_key_prefix(values)));
+        } else if (auto dv = dynamic_pointer_cast<lists::delayed_value>(binop.rhs)) {
+            must_be_in_operator(binop);
+            process_in_values(
+                    dv->get_elements() | transformed(
+                            [&] (const ::shared_ptr<term>& t) {
+                                return static_pointer_cast<tuples::value>(t->bind(options))->get_elements();
+                            }));
+        } else if (auto mkr = dynamic_pointer_cast<tuples::in_marker>(binop.rhs)) {
+            // This is `(a,b) IN ?`.  RHS elements are themselves tuples, represented as vector<bytes_opt>.
+            must_be_in_operator(binop);
+            process_in_values(
+                    static_pointer_cast<tuples::in_value>(mkr->bind(options))->get_split_values());
+        }
+    }
+
+    void operator()(const conjunction& c) {
+        std::ranges::for_each(c.children, [this] (const expression& child) { std::visit(*this, child); });
+    }
+
+    void operator()(bool b) {
+        if (!b) {
+            ranges.clear();
+        }
+    }
+
+    static void must_be_in_operator(const binary_operator& binop) {
+        if (binop.op != oper_t::IN) {
+            on_internal_error(rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
```

Comment:
Where do we verify this?

Reply:
(if this is the first time, it's not an internal error but a user error).

Reply:
> Where do we verify this?

In the grammar.  The parser generates certain terms for IN and others for comparisons.

Reply:
Ok.